### PR TITLE
explorer: add decode `memo` to ascii utf8

### LINKED ISSUE #2527: explorer: Decode memo to ASCII/UTF8
#### Summary
Memos are currently displayed as HEX, unintelligible for the average user. It would be cool if users could check the meaning of memos.

#### Possible solution design or implementation
We could add a decode option that tries to convert the HEX value to ASCII/UTF8.

### PR DESCRIPTION
**Pull request by deuch13**
Resolves #2527

### CODE REVIEW

File: explorer/src/lib/components/transaction-details/TransactionDetails.css [hunk 1]
```diff
@@ -38,7 +61,12 @@
     grid-template-columns: minmax(0, auto) minmax(0, 3fr);
   }
 
-  .transaction-details__payload {
+  .transaction-details__memo {
+    margin-bottom: 1rem;
```
Comment (nortonandreev) [inline explorer/src/lib/components/transaction-details/TransactionDetails.css:65]: Why is that?
Screenshot 2024-10-02 at 13 07 36

It looks better to me if aligned?

Screenshot 2024-10-02 at 13 07 46

File: explorer/src/lib/components/transaction-details/TransactionDetails.svelte [hunk 2]
```diff
@@ -53,8 +54,47 @@
...
   $: classes = makeClassName(["transaction-details", className]);
 
+  /**
+   * @param {string} hex
+   * @returns {string}
+   */
+  function checkHexString(hex) {
+    if (!isHex(hex)) {
+      return hex;
+    }
```
Comment (nortonandreev) [inline explorer/src/lib/components/transaction-details/TransactionDetails.svelte:69]: Is this needed at all? Can't we just use the `isHex` function directly. 👀

File: explorer/src/lib/components/transaction-details/TransactionDetails.svelte [hunk 3]
```diff
@@ -53,8 +54,47 @@
...
+  /** @type {boolean} */
+  let isMemoDecoded = false;
+
   $: classes = makeClassName(["transaction-details", className]);
 
+  /**
+   * @param {string} hex
+   * @returns {string}
+   */
+  function checkHexString(hex) {
```
Comment (nortonandreev) [inline explorer/src/lib/components/transaction-details/TransactionDetails.svelte:66]: I would rename this to `decodeHexString`, instead of check.

File: explorer/src/lib/components/transaction-details/TransactionDetails.svelte [hunk 4]
```diff
@@ -53,8 +54,47 @@
...
+      }
+      return bytes;
+    };
+
+    const bytes = hexToBytes();
+    const decoder = new TextDecoder();
+
+    //Check whether the string contains almost all visible Unicode characters up to the end of the BMP
+    /* eslint-disable-next-line no-control-regex */
+    if (/^[\u0000-\u007F\u00A0-\uFFFF]*$/.test(decoder.decode(bytes))) {
```
Comment (nortonandreev) [inline explorer/src/lib/components/transaction-details/TransactionDetails.svelte:87]: You're decoding twice – once to check the regex and then within the `try`. You could optimize things a bit and only decode once and then store the result.

Something like:

```
  let decodedString;
  try {
    decodedString = decoder.decode(bytes);
  } catch {
    return hexString; // If decoding fails, return the original hex string.
  }
```

File: explorer/src/lib/components/transaction-details/TransactionDetails.svelte [hunk 5]
```diff
@@ -53,8 +54,47 @@
...
+   */
+  function checkHexString(hex) {
+    if (!isHex(hex)) {
+      return hex;
+    }
+
+    /**
+     * @returns {Uint8Array}
+     */
+    const hexToBytes = () => {
```
Comment (nortonandreev) [inline explorer/src/lib/components/transaction-details/TransactionDetails.svelte:74]: Generally, I would extract as a helper on its own, so as that it can be tested properly. 

I wanted to see cases when the hex starts with -0x|0x, I am not sure if these has to be stripped first.

File: explorer/src/lib/components/transaction-details/TransactionDetails.css (hunk 1 above)
Comment (deuch13) [reply explorer/src/lib/components/transaction-details/TransactionDetails.css:65]: Did not notice the misalignment there, the margin is needed to create space between the decoded memo and opened payload.

File: explorer/src/lib/dusk/string/__tests__/isValidHex.spec.js
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/__tests__/isValidHex.spec.js:1]: Lovely!

Some more cases to consider, just for completeness:

1. Test for uppercase hex
`expect(isValidHex("A1B2")).toBe(true);`

2. An edge case
```
it("should return false for a string with only the '0x' prefix", () => {
    expect(isValidHex("0x")).toBe(false);
});
```

3. Whitespace handling
```
it("should return false for strings with leading or trailing spaces", () => {
  expect(isValidHex(" 0x1a2b ")).toBe(false);
});
```

File: explorer/src/lib/dusk/string/decodeHexString.js
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/decodeHexString.js:1]: Few suggestions here – follow in-line comments

File: explorer/src/lib/dusk/string/decodeHexString.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Decodes a hexadecimal string to utf-8 and formats it if it's a JSON string
+ *
+ * @param {string} value
+ * @returns {string}
+ */
+function decodeHexString(value) {
+  if (!isValidHex(value)) {
+    return value;
```
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/decodeHexString.js:11]: Do we want to return the value/string or should we throw an error instead? 

Returning the value directly could be confusing to the caller since the function's purpose is to decode the hex string.

File: explorer/src/lib/dusk/string/decodeHexString.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  const bytes = hexToBytes(value);
+  const decoder = new TextDecoder();
+  const decodedString = decoder.decode(bytes);
```
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/decodeHexString.js:16]: This could fail, if the byte sequence does not form a valid UTF-8 string.

We could wrap in try/catch block - if you want to return null in case of failure or throw the error is up to you.

File: explorer/src/lib/dusk/string/decodeHexString.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+  //Check whether the string contains almost all visible Unicode characters up to the end of the BMP
+  /* eslint-disable-next-line no-control-regex */
+  if (/^[\u0000-\u007F\u00A0-\uFFFF]*$/.test(decodedString)) {
```
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/decodeHexString.js:20]: Perhaps the regex here is too broad – it allows control and invisible characters. Perhaps we could be more restrictive and go with something like `/^[\u0020-\u007E\u00A0-\u00FF]*$/` instead.

File: explorer/src/lib/dusk/string/decodeHexString.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    try {
+      return JSON.stringify(JSON.parse(decodedString), null, 2);
```
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/decodeHexString.js:22]: A suggestion for better readability: 
```
const json = JSON.parse(decodedString);
return JSON.stringify(json, null, 2);
```

File: explorer/src/lib/dusk/string/__tests__/decodeHexString.spec.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  it("should return the input value if it's not a valid hex string", () => {
+    expect(decodeHexString("invalidHex")).toBe("invalidHex");
+  });
+
+  it("should return the decoded string even if it's not valid JSON", () => {
+    expect(decodeHexString("48656c6c6f21")).toBe("Hello!");
+  });
+
+  it("should return the input value if decoded string contains invalid characters", () => {
+    expect(decodeHexString("F09F9880")).toBe("F09F9880");
```
Comment (nortonandreev) [inline explorer/src/lib/dusk/string/__tests__/decodeHexString.spec.js:45]: This could be a bit confusing, as `F09F9880` is valid (it's an emoji).

**Review by nortonandreev — APPROVED** [review]
Great job! 🚀

I will approve the PR, as I don't see anything really blocking. However, I have left a few suggestions here and there. Feel free to incorporate anything you feel is useful.

File: explorer/src/lib/dusk/string/decodeHexString.js (hunk 6 above)
Comment (deuch13) [reply explorer/src/lib/dusk/string/decodeHexString.js:11]: Because the memo can be anything, I do not want the function to throw and error when called but still display the initial value.

File: explorer/src/lib/dusk/string/__tests__/decodeHexString.spec.js (hunk 10 above)
Comment (deuch13) [reply explorer/src/lib/dusk/string/__tests__/decodeHexString.spec.js:45]: Yes, but that emoji unicode is not in this range `\u0000-\u007F\u00A0-\uFFFF` so essentially it is valid

**Review by nortonandreev — APPROVED** [review]
